source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -606,6 +677,85 @@ void Filter::onNewTimeout(const ProtobufWkt::Duration& override_message_timeout)
+    } else if (const auto metadata_it = request_encoder_metadata.find(context_key);
+               metadata_it != request_encoder_metadata.end()) {
+      (*metadata_context.mutable_filter_metadata())[metadata_it->first] = metadata_it->second;
+    } else if (const auto metadata_it = connection_decoder_metadata.find(context_key);
+               metadata_it != connection_decoder_metadata.end()) {
+      (*metadata_context.mutable_filter_metadata())[metadata_it->first] = metadata_it->second;
+    } else if (const auto metadata_it = connection_encoder_metadata.find(context_key);
+               metadata_it != connection_encoder_metadata.end()) {
+      (*metadata_context.mutable_filter_metadata())[metadata_it->first] = metadata_it->second;
+    }
+  }
+
+  // If typed_metadata_context_namespaces is specified, pass matching typed filter metadata to the
+  // ext_proc service. If metadata key is set in both the connection and request metadata then
+  // the value will be the request metadata value. If metadata key is set in both the encoder and
+  // decoder streams, then the value will be the encoder value
+  const auto& connection_decoder_typed_metadata =
+      decoder_callbacks_->connection()->streamInfo().dynamicMetadata().typed_filter_metadata();
+  const auto& request_decoder_typed_metadata =
+      decoder_callbacks_->streamInfo().dynamicMetadata().typed_filter_metadata();
+  const auto& connection_encoder_typed_metadata =
+      encoder_callbacks_->connection()->streamInfo().dynamicMetadata().typed_filter_metadata();
+  const auto& request_encoder_typed_metadata =
+      encoder_callbacks_->streamInfo().dynamicMetadata().typed_filter_metadata();
+  for (const auto& context_key : config_->typedMetadataContextNamespaces()) {
+    if (const auto metadata_it = request_decoder_typed_metadata.find(context_key);
+        metadata_it != request_decoder_typed_metadata.end()) {
+      (*metadata_context.mutable_typed_filter_metadata())[metadata_it->first] = metadata_it->second;
+    } else if (const auto metadata_it = request_encoder_typed_metadata.find(context_key);
+               metadata_it != request_encoder_typed_metadata.end()) {
+      (*metadata_context.mutable_typed_filter_metadata())[metadata_it->first] = metadata_it->second;
+    } else if (const auto metadata_it = connection_decoder_typed_metadata.find(context_key);
+               metadata_it != connection_decoder_typed_metadata.end()) {
+      (*metadata_context.mutable_typed_filter_metadata())[metadata_it->first] = metadata_it->second;
+    } else if (const auto metadata_it = connection_encoder_typed_metadata.find(context_key);
+               metadata_it != connection_encoder_typed_metadata.end()) {
+      (*metadata_context.mutable_typed_filter_metadata())[metadata_it->first] = metadata_it->second;
+    }
+  }
+
+  *req.mutable_metadata_context() = metadata_context;
+}
+
+void setDynamicMetadata(std::string ns, Http::StreamFilterCallbacks* cb,
```

Comment:
This assumes we fully trust whatever is sent from the external server, can we add knobs to control (disable, filter ) what could be added into Envoy?

Reply:
I don't think this is necessary for a few reasons:

1. The external processing filter is listed as unknown security posture, only to be used with both trusted upstreams and downstreams
2. The external processing server has nearly arbitrary control over the request/response as it is, so I think it is safe to assume that it should be a trusted piece of code, at least as trusted as a custom Envoy filter which would be able to write dynamic metadata
3. The returned metadata is only able to be written into dynamic metadata in the ext_proc filter's own namespace.

Reply:
1. its going to change soon. We have been vetting it and @yanjunxiang-google is going to flip this to trusted filter soon
2. Not true.. if you look at all the knobs we have added recently, we are really tying to limit what an external server can and cannot do so that those who want to use it against trusted systems can do so without knobs but those who need to use it against untrusted systems can rely on the knobs to tightly control what can/cannot be done.
3. follows 2 above in case of untrusted servers

Reply:
I've added the ability to disable writing the returned metadata. I think this is a good level of control for this initial implementation, but if more granularity is required for this work to merge, can you please point me to another similar implementation of the filtering?

Reply:
Thanks for adding the control knob. Yes, that's right.  We are going to flip this filter to be robust to untrusted downstream/upstream soon.